components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/wireguard_tunnel_service.cc

```diff
@@ -0,0 +1,290 @@
+bool CreateAndRunBraveWireguardService(const std::wstring& encoded_config) {
+  std::string config;
+  if (!base::Base64Decode(base::WideToUTF8(encoded_config), &config) ||
+      config.empty()) {
+    VLOG(1) << "Unable to decode wireguard config";
+    return S_FALSE;
+  }
+  auto config_file_path = WriteConfigToFile(config);
+  if (!config_file_path.has_value()) {
+    VLOG(1) << "Unable to save config file";
+    return false;
+  }
+
+  ScopedScHandle scm(::OpenSCManager(nullptr, nullptr, SC_MANAGER_ALL_ACCESS));
+  if (!scm.IsValid()) {
+    VLOG(1) << "::OpenSCManager failed. service_name: "
+            << GetBraveVpnWireguardTunnelServiceName()
+            << ", error: " << std::hex << HRESULTFromLastError();
+    return false;
+  }
+
+  base::FilePath directory;
+  if (!base::PathService::Get(base::DIR_EXE, &directory)) {
+    return false;
+  }
+  base::CommandLine service_cmd(
+      directory.Append(brave_vpn::kBraveVpnWireguardServiceExecutable));
+  service_cmd.AppendSwitchPath(
+      brave_vpn::kBraveVpnWireguardServiceConnectSwitchName,
+      config_file_path.value());
+
+  ScopedScHandle service(::CreateService(
+      scm.Get(), GetBraveVpnWireguardTunnelServiceName().c_str(),
+      GetBraveVpnWireguardTunnelServiceName().c_str(), SERVICE_ALL_ACCESS,
+      SERVICE_WIN32_OWN_PROCESS, SERVICE_DEMAND_START, SERVICE_ERROR_NORMAL,
+      service_cmd.GetCommandLineString().c_str(), NULL, NULL, L"Nsi\0TcpIp\0",
+      NULL, NULL));
+  if (!service.IsValid()) {
+    VLOG(1) << "::CreateService failed. service_name: "
+            << GetBraveVpnWireguardTunnelServiceName() << ", error: 0x"
+            << ::GetLastError();
+    return false;
+  }
+
+  SERVICE_SID_INFO info;
+  info.dwServiceSidType = SERVICE_SID_TYPE_UNRESTRICTED;
+  if (!ChangeServiceConfig2(service.Get(), SERVICE_CONFIG_SERVICE_SID_INFO,
+                            &info)) {
+    VLOG(1) << "Failed to configure service 0x" << std::hex
+            << HRESULTFromLastError();
+    return false;
+  }
+
+  if (!brave_vpn::SetServiceFailureActions(service.Get())) {
+    VLOG(1) << "SetServiceFailActions failed:" << std::hex
+            << HRESULTFromLastError();
+    return false;
+  }
+
+  if (!StartService(service.Get(), 0, NULL)) {
+    VLOG(1) << "Failed to start service 0x" << std::hex
+            << HRESULTFromLastError();
+    return false;
+  }
+
+  return DeleteService(service.Get()) != 0;
+}
+
+int RunWireguardTunnelService(const base::FilePath& config_file_path) {
+  base::ScopedTempDir config_dir;
+  if (config_file_path.empty() || !config_dir.Set(config_file_path.DirName())) {
+    VLOG(1) << "Wrong path to config file:" << config_file_path;
+    return S_FALSE;
+  }
+
+  {
+    base::FilePath directory;
+    if (!base::PathService::Get(base::DIR_EXE, &directory)) {
+      return S_FALSE;
+    }
+    typedef bool WireGuardTunnelService(const LPCWSTR settings);
+    base::ScopedNativeLibrary tunnel_lib(directory.Append(L"tunnel.dll"));
```

Comment:
NOTE: this dependency is not added in this PR

We are working out the compilation of that dependency separately 😄

Reply:
yes, this also mentioned in the description